src/cpp/src/image_generation/flux_fill_pipeline.hpp

```diff
@@ -0,0 +1,267 @@
+            m_perf_metrics.raw_metrics.iteration_durations.emplace_back(MicroSeconds(step_ms));
+        }
+
+        latents = unpack_latents(latents, m_custom_generation_config.height, m_custom_generation_config.width, vae_scale_factor);
+        const auto decode_start = std::chrono::steady_clock::now();
+        auto image = m_vae->decode(latents);
+        m_perf_metrics.vae_decoder_inference_duration =
+            std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - decode_start)
+                .count();
+        m_perf_metrics.generate_duration =
+            std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - gen_start).count();
+        return image;
+    }
+
+private:
+
+void check_image_size(const int height, const int width) const override {
+    assert(m_transformer != nullptr);
+    const size_t vae_scale_factor = m_vae->get_vae_scale_factor() * 2;
+    OPENVINO_ASSERT((height % vae_scale_factor == 0 || height < 0) && (width % vae_scale_factor == 0 || width < 0),
+                    "Both 'width' and 'height' must be divisible by ",
+                    vae_scale_factor);
+}
+
+void check_inputs(const ImageGenerationConfig& generation_config, ov::Tensor initial_image) const override {
+    OPENVINO_ASSERT(m_pipeline_type == PipelineType::INPAINTING, "FluxFillPipeline supports inpainting mode only");
+
+    check_image_size(generation_config.width, generation_config.height);
+
+    OPENVINO_ASSERT(generation_config.max_sequence_length <= 512, "T5's 'max_sequence_length' must be less or equal to 512");
+    OPENVINO_ASSERT(generation_config.negative_prompt == std::nullopt, "Negative prompt is not used by FluxFillPipeline");
+    OPENVINO_ASSERT(generation_config.negative_prompt_2 == std::nullopt, "Negative prompt 2 is not used by FluxFillPipeline");
+    OPENVINO_ASSERT(generation_config.negative_prompt_3 == std::nullopt, "Negative prompt 3 is not used by FluxFillPipeline");
+    OPENVINO_ASSERT(generation_config.prompt_3 == std::nullopt, "Prompt 3 is not used by FluxFillPipeline");
+}
+
+void transform_mask(ov::Tensor& mask, size_t batch_size, size_t height, size_t width, size_t vae_scale_factor) {
+    ov::Shape mask_shape = mask.get_shape();
+    OPENVINO_ASSERT(mask_shape.size() == 4 && mask_shape[1] == 1, "Unexpected mask shape");
+
+    // Permutation to (0, 2, 4, 1, 3)
+    auto transpose = [&](float* src, float* dst) {
+        for (size_t b = 0; b < batch_size; ++b) {
+            for (size_t h = 0; h < height; ++h) {
+                for (size_t w = 0; w < width; ++w) {
+                    for (size_t vh = 0; vh < vae_scale_factor; ++vh) {
+                        for (size_t vw = 0; vw < vae_scale_factor; ++vw) {
+                            size_t src_idx = b * (height * vae_scale_factor * width * vae_scale_factor) + h * (vae_scale_factor * width * vae_scale_factor) + vh * (width * vae_scale_factor) + w * vae_scale_factor + vw;
```

Comment:
I think some parts of this formula can be computed just once for each `vw`. For example, you can move out of this loop 
`b * (height * vae_scale_factor * width * vae_scale_factor) + h * (vae_scale_factor * width * vae_scale_factor) + vh * (width * vae_scale_factor) + w * vae_scale_factor`. The same trick can be applied for loops above:)

Reply:
compilers can optimize reparative blocks of code inside the loop and compute them only once